engine/src/conversion/analysis/fun/mod.rs

```diff
@@ -246,6 +262,23 @@ pub(crate) struct FnAnalyzer<'a> {
     existing_superclass_trait_api_names: HashSet<QualifiedName>,
 }
 
+/// Leaves the [`FnStructAnalysis::constructors`] at its default for [`add_missing_constructors`]
+/// to fill out.
```

Comment:
I haven't fully traced this, but from your comment this doesn't sound ideal to me.
The whole point of the `AnalysisPhase` arrangement is that fields don't exist until they're actually needed, so they are _compile-time_ guaranteed to be correct at all times. Idiots like me simply cannot make the mistake of relying on a field before it's populated :)
I would therefore prefer that you add an extra `AnalysisPhase` if necessary, which adds the `PublicConstructors` field at the point it's populated.

Reply:
I considered that, but the reasoning in the existing comment on `find_missing_constructors` (renamed to `find_constructors_present` in the latest commit to reflect its new purpose) still seems to apply:
```rust
/// It is tempting to make this a separate analysis phase, to be run later than
/// the function analysis; but that would make the code much more complex as it
/// would need to output a `FnAnalysisBody`. By running it as part of this phase
/// we can simply generate the sort of thing bindgen generates, then ask
/// the existing code in this phase to figure out what to do with it.
```

`find_explicit_items` relies on the analysis from `analyze_foreign_fn_and_subclasses`, and the result of `find_constructors_present` is used by `add_constructors_present` in `FnAnalyzer`. Splitting that up gets complicated. Maybe moving `analyze_foreign_fn_and_subclasses` (and everything it uses) somewhere it can be used both by `FnAnalyzer` and a new `FnWithConstructorsAnalyzer` would work? `FnAnalyzer` doesn't need `extra_apis`, those can all be added in `FnWithConstructorsAnalyzer`. But I think it still ends up kind of tangled together. Thoughts?

I can take a stab at splitting it up tomorrow if you think it'll end up better in the end.

Reply:
Ah fair enough. If that comment still applies then the current arrangement sounds good to me.